x/ccv/provider/keeper/throttle.go

```diff
@@ -0,0 +1,400 @@
+// and then handles any trailing vsc matured packets in that (consumer chain specific) queue.
+//
+// Note: Any packet data which is handled in this method is also deleted from the (consumer chain specific) queue.
+func (k Keeper) HandlePacketDataForChain(ctx sdktypes.Context, consumerChainID string,
+	slashPacketHandler func(sdktypes.Context, string, ccvtypes.SlashPacketData) (bool, error),
+	vscMaturedPacketHandler func(sdktypes.Context, string, ccvtypes.VSCMaturedPacketData),
+) {
+
+	// Store ibc sequence numbers to delete data after iteration is completed
+	seqNums := []uint64{}
+
+	idx := 0
+	k.IteratePendingPacketData(ctx, consumerChainID, func(ibcSeqNum uint64, data interface{}) (stop bool) {
+
+		switch data := data.(type) {
+
+		case ccvtypes.SlashPacketData:
+			if idx == 0 {
+				_, err := slashPacketHandler(ctx, consumerChainID, data)
+				if err != nil {
+					panic(fmt.Sprintf("failed to handle slash packet: %s", err))
+				}
+			} else {
+				// Break iteration, since we've already handled one slash packet
+				stop = true
+				return stop
+			}
+		case ccvtypes.VSCMaturedPacketData:
+			if idx == 0 {
+				panic("data is corrupt, first data struct in queue should be slash packet data")
+			} else {
+				vscMaturedPacketHandler(ctx, consumerChainID, data)
+			}
+		default:
+			panic(fmt.Sprintf("unexpected pending packet data type: %T", data))
+		}
+		seqNums = append(seqNums, ibcSeqNum)
+
+		// Increment idx
+		idx++
+
+		// Continue iterating through the queue until we reach the end or a 2nd slash packet
+		stop = false
+		return stop
+	})
+
+	// Delete handled data after iteration is completed
+	k.DeletePendingPacketData(ctx, consumerChainID, seqNums...)
+}
+
+// InitializeSlashMeter initializes the slash meter to it's max value (also its allowance
+// per replenish period), and sets the last replenish time to the current block time.
+func (k Keeper) InitializeSlashMeter(ctx sdktypes.Context) {
+	k.SetSlashMeter(ctx, k.GetSlashMeterAllowance(ctx))
+	k.SetLastSlashMeterReplenishTime(ctx, ctx.BlockTime())
+}
+
+// CheckForSlashMeterReplenishment checks if the slash gas meter should be replenished, and if so, replenishes it.
+// Note: initial slash meter replenish time is set in InitGenesis
+func (k Keeper) CheckForSlashMeterReplenishment(ctx sdktypes.Context) {
+	lastReplenishTime := k.GetLastSlashMeterReplenishTime(ctx)
+	replenishPeriod := k.GetSlashMeterReplenishPeriod(ctx)
+	if ctx.BlockTime().UTC().After(lastReplenishTime.Add(replenishPeriod)) {
+		k.ReplenishSlashMeter(ctx)
+	}
+}
+
+func (k Keeper) ReplenishSlashMeter(ctx sdktypes.Context) {
+
+	meter := k.GetSlashMeter(ctx)
+	allowance := k.GetSlashMeterAllowance(ctx)
+
+	// Replenish gas up to gas allowance per period. That is, if meter was negative
+	// before being replenished, it'll gain some additional gas. However, if the meter
+	// was 0 or positive in value, it'll be replenished only up to it's allowance for the period.
```

Comment:
The allowance in this context is how much value can be added to the meter per replenish period, it's `allowance = SlashMeterReplenishFrac * totalVotingPower`

Reply:
That makes it sound like the allowance will change depending on time even if nothing else changes (ie. there are no slash packets at all)

Reply:
To clarify this, I've added 

> However, if the meter was 0 or positive in value, it'll be replenished only up to it's allowance for the current block.

Since allowance can change between blocks, depending on the total voting power for the block